mtx/loaderinfo.1
.\" tapeinfo.1  Document copyright 2000 Eric Lee Green
.\"  Program Copyright 2000 Eric Lee Green <eric@badtux.org>
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139,
.\" USA.
.\"
.TH LOADERINFO 1 LOADERINFO1.0
.SH NAME
loaderinfo \- report SCSI tape device info
.SH SYNOPSIS
loaderinfo -f <scsi-generic-device>
.SH DESCRIPTION
The
.B loaderinfo
command reads various information from SCSI tape loaders. Its intended
use is for high-level programs that are trying to decide what the
capabilities of a device are.
.P
The following are printed:
.TP 10
.B  Element Address Assignment Page:
This tells how many elements are in the loader, and what their raw
hardware addresses are.

.TP 10
.B Transport Geometry Descriptor Page:
Will display whether media is invertible or not (usable with some
optical jukeboxes for detirmining whether to "flip" media after writing
to the first side).

.TP 10
.B Device Capabilities Page
Currently will only display whether we can transfer between slots (i.e.
whether 'mtx transfer' works).

.TP 10
.B Inquiry Page
Aside from the normal inquiry info, will also print out whether we have
a bar code reader (for loaders that support the Exabyte extension for
reporting presense of said reader).


.SH OPTIONS
The first argument, given following
.B -f
, is the SCSI generic device corresponding to your tape loader.
Consult your operating system's documentation for more information (for
example, under Linux these are generally start at /dev/sg0
under FreeBSD these start at /dev/pass0).
.P
Under FreeBSD, 'camcontrol devlist' will tell you what SCSI devices you
have, along with which 'pass' device controls them. Under Linux,
the command
.B lsscsi \-\-generic
will tell you what SCSI devices you have.
Under Solaris 8,
.B find /devices -name '*changer*'
will display the device names for your attached changers. Make sure
to configure your 'sgen' driver first.

.SH BUGS AND LIMITATIONS
.P
This program has only been tested on Linux with a limited number of
loaders (Ecrix Autopack, Exabyte 220, IBM TS3100).
.P
.SH AVAILABILITY
.B loaderinfo
is currently being maintained by Robert Nelson <robertnelson@users.sourceforge.net>
as part of the 'mtx' suite of programs. The 'mtx' home page is
http://mtx.sourceforge.net and the actual code is currently available there and via
SVN from http://sourceforge.net/projects/mtx.

.SH SEE ALSO
.BR mt (1), tapeinfo (1), mtx (1)


mtx/CHANGES
CHANGES, mtx 1.3.11
- loaderinfo: Print RequestSense information if an operation fails.
- loaderinfo: Print all information from the Device Configuration Page.
- mtx: Fix problems with a lot of loaders determining the number of elements.
- mtx: Implement Previous command.
- mtx: Fix bugs in Last command.
- scsieject: New tool that handles start/stop, load/unload.
- tapeinfo: Fix inconsistencies in output.
- Fix timeout handling on Linux.
- Merge in fixes from the Debian and FreeBSD distributions.

CHANGES, mtx 1.3.10
- Add support for Microsoft Windows.
- Add support for Sony VGP-XL1B(2) Media Changer.
  Thanks to Will (nodenet at hotmail dot com)
- Add build support for Microsoft Windows using MinGW native and Linux
  cross-compile.
- Add build support for Microsoft Windows using Microsoft Visual Studio 2005.
- Add support for building on x86_64.
- Add more debugging information.
- Eliminate compiler warnings.

CHANGES, mtx 1.3.9
- Cleaned up scsi_linux.c a little to eliminate around 40 lines of code. 
- Change to using SG_IO ioctl rather than write-read interface. This
  should make us a little safer, as well as (on 2.6.x) letting us issue SCSI 
  commands to regular block devices as well as to /dev/sgXX devices. 
  WARNING: Can cause the system to *CRASH* if the SCSI device is in use, due 
  to brokenness inside the Linux kernel! It's always preferable to use the 
  /dev/sgXX device, which has specialness in its buffer handling that bypasses
  some of the brokenness elsewhere in the SCSI subsystem.
- Check SG version before using SG_IO interface, so that if we're run on a
  Linux 2.2 system and we were compiled on Linux 2.4 or above, 
- If a tape is in the drive, make tapeinfo print out its partition info. Seems
  to work on my DAT drive, anyhow (shrug). 
- Minor cleanups in mtxl.c (other cleanups necessary in tapeinfo and etc.). 

CHANGES, mtx 1.3.8
- Forgot about 8 bytes for header
- Handle pedantic loader whose firmware writer spends too much time
  parsing phrases in the SCSI spec and not enough time in the
  real world. 

CHANGES, mtx 1.3.6
- The You Gotta Be F'ing Kidding Release (see rant on mailing list)
- Added AIX support with http://fz.eryx.net/aix/ GSC driver, courtesy
  of Steve Heck. 
- Make __WEIRD_CHAR_SUPPRESS the default to stop barcode garbling
- Fix core dump on invalid inputs in mtx.c
- Add retry with bigger ALLOCATION_LENGTH if 
  BYTE_COUNT_OF_REPORT_AVAILABLE is > than our original estimate. (see rant
  on mailing list). 
- Shut up the low-level SCSI sg_read and sg_write routines, which were poofting
  all over the place on innocuous things like no EAAP. 

CHANGES, mtx 1.3.5
- The I-Need-A-Job Release (see http://badtux.org) :-}
- Change Makefile so that it works w/systems whose 'install' program
  accepts only one argument. That's it. 
- Uncomment the Exabyte hack in the drive parsing code so that it works
  with ATL loaders again. 
- Add hack for ATL stacked loaders to keep them from reporting ghost slots

CHANGES, mtx 1.3.4
- Forward-ported the ATL/Exabyte patches from mtx 1.2
- Added progress indicator to 'tapeinfo' for DAT drives. (PLEASE TEST)
- Purged a few more estinc.com out of the documentation, point to bru.com.

CHANGES, mtx 1.3.3
- Reversed James' long descriptor patch until can figure out what's wrong with
  it.
- Fixed some debug declarations, courtesy of Kevin Wang. 
- Point COMPATIBILITY file at mtx.sourceforge.net.

CHANGES, mtx 1.3.2
- Merged Makefile changes from mtx-1.2 code
- Changes from James Dugal to get all info for newer loaders, fix debugging

CHANGES, mtx 1.3.1
- add an Exchange command

CHANGES, mtx 1.3.0
- New Linux SCSI subsystem interface for 2.4 kernels, ripped boldly 
  from sgtools by Doug Gilbert
- mtx 'next' now skips blank slots, courtesy of Christopher McCrory
- mtx 'unload' now prints an error message telling you that you need to
  eject the tape in the drive prior to unloading it, under certain
  conditions. 
- Started work on 'nsmhack' for controlling NSM jukeboxes. 
- copy_barcode was off by one. 
- 'position to element' command now added, courtesy of Mahlon Stacy

CHANGES, mtx 1.2.15:
- Some Solaris fixes, courtesy of Matt Ward
- Fix URL in .spec file

CHANGES, mtx 1.2.14:
- Fix so it'll work if 0 is result of SCSI open (e.g., in cron jobs on Linux)
- Move changelog to end of .spec file for easier reading
- Added a bit of text to beginning of COMPATIBILITY file

CHANGES, mtx 1.2.13:
- Fixed some autoloader bugs w/autoloaders that don't report an arm. 
- Fixed barcode backoff.
- Added "nobarcode" option
- Increased timeout for 'mtx inventory' to 30 minutes
  (note: may increase this even more if needed, please let me know!)
- Shortened timeout for 'mtx inquiry' to 30 seconds
- tapeinfo now prints SCSI ID/LUN info if available (only on Linux at the
  moment, sigh). 
- update documentation w/new email addresses, updated compile directions,
  various fixes. 

CHANGES, mtx 1.2.12:
- Fix FreeBSD compile bugs
- Fix SGI compile bugs
- Add HP/UX port (I hope!), courtesy of William Smith. 
- Re-wrote ReadElementStatus to make work for %!@# brain dead firmware that
  reports non-existent drives (I hope!). Also has side-effect of now working
  with multiple-arm libraries (though it only sees first arm!).
- Cleaned up all -Wall messages.
- Cleaned up Linux Sparc, installs loaderinfo.1, courtesy of Matt Dainty.
- tapeinfo now reports status of CheckUnitReady.  
- tapeinfo no longer puts out Block Position if CheckUnitReady says 'no'.
- tapeinfo now puts out Density Code and Medium Type/Not Loaded (modification
    of patches sent in by Bob Rahe)

CHANGES, mtx 1.2.11:
- Added a GNU autoconf Makefile.in (still provide a Makefile for your use)
  *WARNING* autoconf is not yet working on all supported OS's! You may need
  to do the old-fashioned 'edit Makefile' bit!
- Changed mtx.h and mtxl.c to include and define various things based upon
  what features autoconf detected (e.g., if 'camlib.h', use FreeBSD-style

mtx/loaderinfo.cpp
/* Copyright 2000 Enhanced Software Technologies Inc.
 *   Released under terms of the GNU General Public License as
 * required by the license on 'mtxl.c'.
 */

/* 
* $Date: 2007-03-26 05:39:27 +0000 (Mon, 26 Mar 2007) $
* $Revision: 188 $
*/

/* What this does: Basically dumps out contents of:
 *  Mode Sense: Element Address Assignment Page (0x1d)
 *           1Eh (Transport Geometry Parameters) has a bit which indicates is 
 *               a robot is  capable of rotating the media. It`s the 
 *                `Rotate` bit, byte 2, bit 1.
 *          Device Capabilities page (0x1f)
 * Inquiry -- prints full inquiry info. 
 *   DeviceType:
 *    Manufacturer:
 *    ProdID:  
 *    ProdRevision:
 *   If there is a byte 55, we use the Exabyte extension to 
 * print out whether we have a bar code reader or not.  This is
 * bit 0 of byte 55. 
 *
 * Next, we request element status on the drives. We do not
 * request volume tags though. If Exabyte
 * extensions are supported, we report the following information for
 * each drive:
 *
 *  Drive number
 *  EXCEPT (with ASC and ASCQ), if there is a problem. 
 *  SCSI address and LUN
 *  Tape drive Serial number
 *   
 */

#include <stdio.h>
#include "mtx.h"
#include "mtxl.h"

DEVICE_TYPE MediumChangerFD;  /* historic purposes... */

char *argv0;

/* A table for printing out the peripheral device type as ASCII. */ 
static char *PeripheralDeviceType[32] =
{
	"Disk Drive",
	"Tape Drive",
	"Printer",
	"Processor",
	"Write-once",
	"CD-ROM",
	"Scanner",
	"Optical",
	"Medium Changer",
	"Communications",
	"ASC IT8",
	"ASC IT8",
	"RAID Array",
	"Enclosure Services",
	"OCR/W",
	"Bridging Expander", /* 0x10 */
	"Reserved",  /* 0x11 */
	"Reserved", /* 0x12 */
	"Reserved",  /* 0x13 */
	"Reserved",  /* 0x14 */
	"Reserved",  /* 0x15 */
	"Reserved",  /* 0x16 */
	"Reserved",  /* 0x17 */
	"Reserved",  /* 0x18 */
	"Reserved",  /* 0x19 */
	"Reserved",  /* 0x1a */
	"Reserved",  /* 0x1b */
	"Reserved",  /* 0x1c */
	"Reserved",  /* 0x1d */
	"Reserved",  /* 0x1e */
	"Unknown"    /* 0x1f */
};


/* okay, now for the structure of an Element Address Assignment Page:  */

typedef struct EAAP
{
	unsigned char Page_Code;
	unsigned char Parameter_Length;
	unsigned char MediumTransportElementAddress[2];
	unsigned char NumMediumTransportElements[2];
	unsigned char FirstStorageElementAdddress[2];
	unsigned char NumStorageElements[2];
	unsigned char FirstImportExportElementAddress[2];
	unsigned char NumImportExportElements[2];
	unsigned char FirstDataTransferElementAddress[2];
	unsigned char NumDataTransferElements[2];
	unsigned char Reserved[2];
}	EAAP_Type;

/* okay, now for the structure of a transport geometry
 * descriptor page:
 */
typedef struct TGDP
{
	unsigned char Page_Code;
	unsigned char ParameterLength;
	unsigned char Rotate;
	unsigned char ElementNumber;  /* we don't care about this... */
}	TGDP_Type;


/* Structure of the Device Capabilities Page: */
typedef struct DCP 
{
	unsigned char Page_Code;
	unsigned char ParameterLength;
	unsigned char CanStore;		/* bits about whether elements can store carts */
	unsigned char SMC2_Caps;
	unsigned char MT_Transfer;	/* bits about whether mt->xx transfers work. */
	unsigned char ST_Transfer;	/* bits about whether st->xx transfers work. */
	unsigned char IE_Transfer;	/* bits about whether id->xx transfers work. */
	unsigned char DT_Transfer;	/* bits about whether DT->xx transfers work. */
	unsigned char Reserved[4];	/* more reserved data */
	unsigned char MT_Exchange;	/* bits about whether mt->xx exchanges work. */
	unsigned char ST_Exchange;	/* bits about whether st->xx exchanges work. */
	unsigned char IE_Exchange;	/* bits about whether id->xx exchanges work. */
	unsigned char DT_Exchange;	/* bits about whether DT->xx exchanges work. */
	unsigned char Reserved2[4];	/* more reserved data */
}	DCP_Type;

#define MT_BIT 0x01
#define ST_BIT 0x02
#define IE_BIT 0x04
#define DT_BIT 0x08

/* Okay, now for the inquiry information: */

static void ReportInquiry(DEVICE_TYPE MediumChangerFD)
{
	RequestSense_T RequestSense;
	Inquiry_T *Inquiry;
	int i;

	Inquiry = RequestInquiry(MediumChangerFD,&RequestSense);
	if (Inquiry == NULL) 
	{
		PrintRequestSense(&RequestSense);
		FatalError("INQUIRY Command Failed\n");
	}

	printf("Product Type: %s\n",PeripheralDeviceType[Inquiry->PeripheralDeviceType]);

	printf("Vendor ID: '");
	for (i = 0; i < sizeof(Inquiry->VendorIdentification); i++)
		printf("%c", Inquiry->VendorIdentification[i]);

	printf("'\nProduct ID: '");
	for (i = 0; i < sizeof(Inquiry->ProductIdentification); i++)
		printf("%c", Inquiry->ProductIdentification[i]);

	printf("'\nRevision: '");
	for (i = 0; i < sizeof(Inquiry->ProductRevisionLevel); i++)
		printf("%c", Inquiry->ProductRevisionLevel[i]);

	printf("'\n");

	if (Inquiry->MChngr) 
	{
		/* check the attached-media-changer bit... */
		printf("Attached Changer: Yes\n");
	}
	else
	{
		printf("Attached Changer: No\n");
	}

	/* Now see if we have a bar code flag: */
	if (Inquiry->AdditionalLength > 50) 
	{
		/* see if we have 56 bytes: */
		if (Inquiry->VendorFlags & 1)
		{
			printf("Bar Code Reader: Yes\n");
		}
		else
		{
			printf("Bar Code Reader: No\n");
		}
	}

	free(Inquiry);		/* well, we're about to exit, but ... */
}

/*********** MODE SENSE *******************/
/* We need 3 different mode sense pages. This is a generic
 * routine for obtaining mode sense pages. 
 */

static unsigned char
*mode_sense(DEVICE_TYPE fd, char pagenum, int alloc_len,  RequestSense_T *RequestSense)
{
	CDB_T CDB;
	unsigned char *input_buffer;	/*the input buffer -- has junk prepended to
									 * actual sense page. 
									 */
	unsigned char *tmp;
	unsigned char *retval;			/* the return value. */
	int i,pagelen;

	if (alloc_len > 255)
	{
		FatalError("mode_sense(6) can only read up to 255 characters!\n");
	}

	input_buffer = (unsigned char *)xzmalloc(256); /* overdo it, eh? */

	/* clear the sense buffer: */
	memset(RequestSense, 0, sizeof(RequestSense_T));

	/* returns an array of bytes in the page, or, if not possible, NULL. */
	CDB[0] = 0x1a; /* Mode Sense(6) */
	CDB[1] = 0x08; 
	CDB[2] = pagenum; /* the page to read. */
	CDB[3] = 0;
	CDB[4] = 255; /* allocation length. This does max of 256 bytes! */
	CDB[5] = 0;

	if (SCSI_ExecuteCommand(fd, Input, &CDB, 6,
							input_buffer, 255, RequestSense) != 0)
	{
#ifdef DEBUG_MODE_SENSE
		fprintf(stderr,"Could not execute mode sense...\n");
		fflush(stderr);
#endif
		return NULL; /* sorry, couldn't do it. */
	}

	/* First skip past any header.... */
	tmp = input_buffer + 4 + input_buffer[3];
	/* now find out real length of page... */
	pagelen = tmp[1] + 2;
	retval = (unsigned char *)xmalloc(pagelen);
	/* and copy our data to the new page. */
	for (i = 0; i < pagelen; i++)
	{
		retval[i] = tmp[i];
	}
	/* okay, free our input buffer: */
	free(input_buffer);
	return retval;
}

/* Report the Element Address Assignment Page */
static void ReportEAAP(DEVICE_TYPE MediumChangerFD)
{
	EAAP_Type *EAAP; 
	RequestSense_T RequestSense;

	EAAP = (EAAP_Type *)mode_sense(MediumChangerFD, 0x1d, sizeof(EAAP_Type), &RequestSense);

	if (EAAP == NULL)
	{
		PrintRequestSense(&RequestSense);
		printf("EAAP: No\n");
		return;
	}

	/* we did get an EAAP, so do our thing: */
	printf("EAAP: Yes\n");
	printf("Number of Medium Transport Elements: %d\n", ( ((unsigned int)EAAP->NumMediumTransportElements[0]<<8) + (unsigned int)EAAP->NumMediumTransportElements[1]));
	printf("Number of Storage Elements: %d\n", ( ((unsigned int)EAAP->NumStorageElements[0]<<8) + (unsigned int)EAAP->NumStorageElements[1]));
	printf("Number of Import/Export Elements: %d\n", ( ((unsigned int)EAAP->NumImportExportElements[0]<<8) + (unsigned int)EAAP->NumImportExportElements[1]));
	printf("Number of Data Transfer Elements: %d\n", ( ((unsigned int)EAAP->NumDataTransferElements[0]<<8) + (unsigned int)EAAP->NumDataTransferElements[1]));

	free(EAAP);
}

/* See if we can get some invert information: */

static void Report_TGDP(DEVICE_TYPE MediumChangerFD)
{
	TGDP_Type *result;

	RequestSense_T RequestSense;

	result=(TGDP_Type *)mode_sense(MediumChangerFD,0x1e,255,&RequestSense);

	if (!result)
	{
		printf("Transport Geometry Descriptor Page: No\n");
		return;
	}

	printf("Transport Geometry Descriptor Page: Yes\n");

	/* Now print out the invert bit: */
	if ( result->Rotate & 1 )
	{
		printf("Invertable: Yes\n");
	}
	else
	{
		printf("Invertable: No\n");
	}

	free(result);
}

/* Okay, let's get the Device Capabilities Page. We don't care
 * about much here, just whether 'mtx transfer' will work (i.e., 
 * ST->ST).
 */

void TransferExchangeTargets(unsigned char ucValue, char *szPrefix)
{
	if (ucValue & DT_BIT)
	{
		printf("%sData Transfer", szPrefix);
	}

	if (ucValue & IE_BIT)
	{
		printf("%s%sImport/Export", ucValue > (IE_BIT | (IE_BIT - 1)) ? ", " : "", szPrefix);
	}

	if (ucValue & ST_BIT)
	{
		printf("%s%sStorage", ucValue > (ST_BIT | (ST_BIT - 1)) ? ", " : "", szPrefix);
	}

	if (ucValue & MT_BIT)
	{
		printf("%s%sMedium Transfer", ucValue  > (MT_BIT | (MT_BIT - 1)) ? ", " : "", szPrefix);
	}
}

static void Report_DCP(DEVICE_TYPE MediumChangerFD)
{
	DCP_Type *result;
	RequestSense_T RequestSense;

	/* Get the page. */
	result=(DCP_Type *)mode_sense(MediumChangerFD,0x1f,sizeof(DCP_Type),&RequestSense);
	if (!result) 
	{
		printf("Device Configuration Page: No\n");
		return;
	}

	printf("Device Configuration Page: Yes\n");

	printf("Storage: ");

	if (result->CanStore & DT_BIT)
	{
		printf("Data Transfer");
	}

	if (result->CanStore & IE_BIT)
	{
		printf("%sImport/Export", result->CanStore > (IE_BIT | (IE_BIT - 1)) ? ", " : "");
	}

	if (result->CanStore & ST_BIT)
	{
		printf("%sStorage", result->CanStore > (ST_BIT | (ST_BIT - 1)) ? ", " : "");
	}

	if (result->CanStore & MT_BIT)
	{
		printf("%sMedium Transfer", result->CanStore > (MT_BIT | (MT_BIT - 1)) ? ", " : "");
	}

	printf("\n");

	printf("SCSI Media Changer (rev 2): ");

	if (result->SMC2_Caps & 0x01)
	{
		printf("Yes\n");

		printf("Volume Tag Reader Present: %s\n", result->SMC2_Caps & 0x02 ? "Yes" : "No");
		printf("Auto-Clean Enabled: %s\n", result->SMC2_Caps & 0x04 ? "Yes" : "No");
	}
	else
	{
		printf("No\n");
	}

	printf("Transfer Medium Transport: ");
	if ((result->MT_Transfer & 0x0F) != 0)
	{
		TransferExchangeTargets(result->MT_Transfer, "->");
	}
	else
	{
		printf("None");
	}

	printf("\nTransfer Storage: ");
	if ((result->ST_Transfer & 0x0F) != 0)
	{
		TransferExchangeTargets(result->ST_Transfer, "->");
	}
	else
	{
		printf("None");
	}

	printf("\nTransfer Import/Export: ");
	if ((result->IE_Transfer & 0x0F) != 0)
	{
		TransferExchangeTargets(result->IE_Transfer, "->");
	}
	else
	{
		printf("None");
	}

	printf("\nTransfer Data Transfer: ");
	if ((result->DT_Transfer & 0x0F) != 0)
	{
		TransferExchangeTargets(result->DT_Transfer, "->");
	}
	else
	{
		printf("None");
	}

	printf("\nExchange Medium Transport: ");
	if ((result->MT_Exchange & 0x0F) != 0)
	{
		TransferExchangeTargets(result->MT_Exchange, "<>");
	}
	else
	{
		printf("None");
	}

	printf("\nExchange Storage: ");
	if ((result->ST_Exchange & 0x0F) != 0)
	{
		TransferExchangeTargets(result->ST_Exchange, "<>");
	}
	else
	{
		printf("None");
	}

	printf("\nExchange Import/Export: ");
	if ((result->IE_Exchange & 0x0F) != 0)
	{
		TransferExchangeTargets(result->IE_Exchange, "<>");
	}
	else
	{
		printf("None");
	}

	printf("\nExchange Data Transfer: ");
	if ((result->DT_Exchange & 0x0F) != 0)
	{
		TransferExchangeTargets(result->DT_Exchange, "<>");
	}
	else
	{
		printf("None");
	}

	printf("\n");

	free(result);
}

void usage(void)
{
	FatalError("Usage: loaderinfo -f <generic-device>\n");
}


/* we only have one argument: "-f <device>". */
int main(int argc, char **argv)
{
	DEVICE_TYPE fd;
	char *filename;

	argv0=argv[0];
	if (argc != 3)
	{
	        if (argc < 3)
                {
                        fprintf(stderr,"ERROR: Too few arguments\n");
                }
                if (argc > 3)
                {
                        fprintf(stderr,"ERROR: %d too many arguments\n", argc-3);
                }
		usage();
	}

	if (strcmp(argv[1],"-f")!=0)
	{
		usage();
	}

	filename=argv[2];

	fd=SCSI_OpenDevice(filename);

	/* Now to call the various routines: */
	ReportInquiry(fd);
	ReportEAAP(fd);
	Report_TGDP(fd);
	Report_DCP(fd);
	exit(0);
}
